File: src/services/timeOffTypeService.ts
/**
 * Time Off Type Service
 * Handles all API interactions for Time Off Types
 * Based on Time_Off_Type.md specification + P6.1 Admin API
 */

import { apiClient } from '@/lib/api';
import { extractApiResponse } from '@/utils/apiResponse';
import {
  TimeOffType,
  TimeOffTypeListResponse,
  CreateTimeOffTypeDto,
  UpdateTimeOffTypeDto,
} from '@/types/timeOff';

export class TimeOffTypeService {
  private static readonly BASE_URL = '/time-off-types';
  private static readonly ADMIN_BASE_URL = '/admin/time-off-types';

  /**
   * ============================================
   * EMPLOYEE APIs (Read-only for creating requests)
   * ============================================
   */

  /**
   * Lấy danh sách loại nghỉ phép (Employee view - chỉ active)
   * 
   * Requires: VIEW_LEAVE_TYPE permission
   * 
   * Query parameters:
   * - isActive: Filter by active status (true/false)
   * - requiresApproval: Filter by approval requirement (true/false)
   * - isPaid: Filter by payment status (true/false)
   * - page: Page number (default: 0)
   * - size: Page size (default: 20)
   * 
   * Possible errors:
   * - 403: Access Denied (missing VIEW_LEAVE_TYPE permission)
   */
  static async getTimeOffTypes(params?: {
    isActive?: boolean;
    requiresApproval?: boolean;
    isPaid?: boolean;
    page?: number;
    size?: number;
  }): Promise<TimeOffType[]> {
    const axios = apiClient.getAxiosInstance();
    const response = await axios.get<TimeOffType[]>(
      this.BASE_URL,
      { params }
    );
    const data = extractApiResponse<TimeOffType[]>(response);
    return Array.isArray(data) ? data : [];
  }

  /**
   * Lấy danh sách loại nghỉ phép cho dropdown (chỉ active types)
   * 
   * Requires: VIEW_LEAVE_TYPE permission
   */
  static async getActiveTimeOffTypes(): Promise<TimeOffType[]> {
    const response = await this.getTimeOffTypes({
      isActive: true
    });
    return response;
  }

  /**
   * Lấy danh sách loại nghỉ phép yêu cầu phê duyệt
   * 
   * Requires: VIEW_LEAVE_TYPE permission
   */
  static async getApprovalRequiredTypes(): Promise<TimeOffType[]> {
    const response = await this.getTimeOffTypes({
      isActive: true,
      requiresApproval: true
    });
    return response;
  }

  /**
   * ============================================
   * ADMIN APIs (Full CRUD for management)
   * ============================================
   */

  /**
   * 1. GET /api/v1/admin/time-off-types
   * Lấy danh sách TẤT CẢ loại nghỉ phép (bao gồm cả inactive)
   * 
   * Requires: VIEW_LEAVE_TYPE permission
   * 
   * Possible errors:
   * - 403: FORBIDDEN - Insufficient permissions
   */
  static async getAllTimeOffTypes(params?: {
    isActive?: boolean;
    isPaid?: boolean;
  }): Promise<TimeOffType[]> {
    const axios = apiClient.getAxiosInstance();
    const response = await axios.get<TimeOffType[]>(
      this.ADMIN_BASE_URL,
      { params }
    );
    const data = extractApiResponse<TimeOffType[]>(response);
    return Array.isArray(data) ? data : [];
  }

  /**
   * 2. POST /api/v1/admin/time-off-types
   * Tạo loại nghỉ phép mới
   * 
   * Requires: MANAGE_LEAVE_TYPE permission
   * 
   * Possible errors:
   * - 400: BAD_REQUEST - Validation errors
   * - 403: FORBIDDEN - Insufficient permissions
   * - 409: CONFLICT - DUPLICATE_TYPE_CODE
   */
  static async createTimeOffType(data: CreateTimeOffTypeDto): Promise<TimeOffType> {
    const axios = apiClient.getAxiosInstance();
    const response = await axios.post<TimeOffType>(
      this.ADMIN_BASE_URL,
      data
    );
    return extractApiResponse<TimeOffType>(response);
  }

  /**
   * 3. PATCH /api/v1/admin/time-off-types/{id}
   * Cập nhật loại nghỉ phép
   * 
   * Requires: MANAGE_LEAVE_TYPE permission
   * 
   * Possible errors:
   * - 400: BAD_REQUEST - Validation errors
   * - 403: FORBIDDEN - Insufficient permissions
   * - 404: NOT_FOUND - TIMEOFF_TYPE_NOT_FOUND
   * - 409: CONFLICT - DUPLICATE_TYPE_CODE
   */
  static async updateTimeOffType(
    typeId: string,
    data: UpdateTimeOffTypeDto
  ): Promise<TimeOffType> {
    const axios = apiClient.getAxiosInstance();
    const response = await axios.patch<TimeOffType>(
      `${this.ADMIN_BASE_URL}/${typeId}`,
      data
    );
    return extractApiResponse<TimeOffType>(response);
  }

  /**
   * 4. DELETE /api/v1/admin/time-off-types/{id}
   * Toggle trạng thái is_active (soft delete)
   * 
   * Requires: MANAGE_LEAVE_TYPE permission
   * 
   * Business Logic:
   * - Đảo ngược is_active (true -> false hoặc false -> true)
   * - Nếu vô hiệu hóa (true -> false), kiểm tra xem có yêu cầu PENDING nào đang dùng không
   * 
   * Possible errors:
   * - 403: FORBIDDEN - Insufficient permissions
   * - 404: NOT_FOUND - TIMEOFF_TYPE_NOT_FOUND
   * - 409: CONFLICT - TIMEOFF_TYPE_IN_USE (có yêu cầu PENDING đang dùng)
   */
  static async deleteTimeOffType(typeId: string): Promise<TimeOffType> {
    const axios = apiClient.getAxiosInstance();
    const response = await axios.delete<TimeOffType>(
      `${this.ADMIN_BASE_URL}/${typeId}`
    );
    return extractApiResponse<TimeOffType>(response);
  }
}

// Export singleton instance
export const timeOffTypeService = new TimeOffTypeService();

File: files_from_BE/working_schedule/service/TimeOffTypeService.java
package com.dental.clinic.management.working_schedule.service;

import com.dental.clinic.management.exception.validation.DuplicateTypeCodeException;
import com.dental.clinic.management.exception.time_off.TimeOffTypeInUseException;
import com.dental.clinic.management.exception.time_off.TimeOffTypeNotFoundException;
import com.dental.clinic.management.exception.validation.BadRequestAlertException;
import com.dental.clinic.management.utils.IdGenerator;
import com.dental.clinic.management.working_schedule.domain.TimeOffType;
import com.dental.clinic.management.working_schedule.dto.request.CreateTimeOffTypeRequest;
import com.dental.clinic.management.working_schedule.dto.request.UpdateTimeOffTypeRequest;
import com.dental.clinic.management.working_schedule.dto.response.TimeOffTypeResponse;
import com.dental.clinic.management.working_schedule.enums.TimeOffStatus;
import com.dental.clinic.management.working_schedule.mapper.TimeOffTypeMapper;
import com.dental.clinic.management.working_schedule.repository.TimeOffRequestRepository;
import com.dental.clinic.management.working_schedule.repository.TimeOffTypeRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Service for managing time-off types
 */
@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(readOnly = true)
public class TimeOffTypeService {

    private final TimeOffTypeRepository typeRepository;
    private final TimeOffRequestRepository requestRepository;
    private final TimeOffTypeMapper typeMapper;
    private final IdGenerator idGenerator;

    /**
     * GET /api/v1/time-off-types
     * Lấy danh sách tất cả các loại hình nghỉ phép đang hoạt động
     */
    public List<TimeOffTypeResponse> getActiveTimeOffTypes() {
        log.debug("Request to get all active time-off types");

        return typeRepository.findByIsActiveTrue()
                .stream()
                .map(typeMapper::toResponse)
                .collect(Collectors.toList());
    }

    /**
     * GET /api/v1/admin/time-off-types
     * Lấy tất cả loại nghỉ phép (bao gồm inactive) - Admin view
     * 
     * @param isActive filter by active status (null = all, true = active only, false = inactive only)
     * @param isPaid filter by paid status (null = all, true = paid only, false = unpaid only)
     */
    public List<TimeOffTypeResponse> getAllTimeOffTypes(Boolean isActive, Boolean isPaid) {
        log.debug("Admin request to get all time-off types, isActive={}, isPaid={}", isActive, isPaid);

        List<TimeOffType> types;
        
        // Apply filters
        if (isActive == null && isPaid == null) {
            // No filters - get all
            types = typeRepository.findAll();
        } else if (isActive != null && isPaid != null) {
            // Both filters
            types = typeRepository.findByIsActiveAndIsPaid(isActive, isPaid);
        } else if (isActive != null) {
            // Only isActive filter
            types = typeRepository.findByIsActive(isActive);
        } else {
            // Only isPaid filter
            types = typeRepository.findByIsPaid(isPaid);
        }

        return types.stream()
                .map(typeMapper::toResponse)
                .collect(Collectors.toList());
    }

    /**
     * GET /api/v1/admin/time-off-types/{typeId}
     * Lấy chi tiết một loại nghỉ phép
     */
    public TimeOffTypeResponse getTimeOffTypeById(String typeId) {
        log.debug("Request to get time-off type by ID: {}", typeId);

        TimeOffType type = typeRepository.findById(typeId)
                .orElseThrow(() -> new TimeOffTypeNotFoundException(typeId));

        return typeMapper.toResponse(type);
    }

    /**
     * POST /api/v1/admin/time-off-types
     * Tạo loại nghỉ phép mới
     */
    @Transactional
    public TimeOffTypeResponse createTimeOffType(CreateTimeOffTypeRequest request) {
        log.debug("Request to create time-off type: {}", request.getTypeCode());

        // Validate unique type_code
        if (typeRepository.existsByTypeCode(request.getTypeCode())) {
            throw new DuplicateTypeCodeException(request.getTypeCode());
        }

        // Validate logic: requiresBalance và defaultDaysPerYear phải match
        validateBalanceAndDefaultDays(request.getRequiresBalance(), request.getDefaultDaysPerYear());

        String typeId = idGenerator.generateId("TOT");

        TimeOffType type = TimeOffType.builder()
                .typeId(typeId)
                .typeCode(request.getTypeCode())
                .typeName(request.getTypeName())
                .description(request.getDescription())
                .requiresBalance(request.getRequiresBalance())
                .defaultDaysPerYear(request.getDefaultDaysPerYear())
                .isPaid(request.getIsPaid())
                .requiresApproval(request.getRequiresApproval() != null ? request.getRequiresApproval() : true)
                .isActive(request.getIsActive() != null ? request.getIsActive() : true)
                .build();

        TimeOffType saved = typeRepository.save(type);
        log.info("Created time-off type: {}", saved.getTypeId());

        return typeMapper.toResponse(saved);
    }

    /**
     * PATCH /api/v1/admin/time-off-types/{typeId}
     * Cập nhật loại nghỉ phép
     */
    @Transactional
    public TimeOffTypeResponse updateTimeOffType(String typeId, UpdateTimeOffTypeRequest request) {
        log.debug("Request to update time-off type: {}", typeId);

        TimeOffType type = typeRepository.findById(typeId)
                .orElseThrow(() -> new TimeOffTypeNotFoundException(typeId));

        // Update type_code if provided and different
        if (request.getTypeCode() != null && !request.getTypeCode().equals(type.getTypeCode())) {
            // Check unique (excluding current type)
            if (typeRepository.existsByTypeCodeAndTypeIdNot(request.getTypeCode(), typeId)) {
                throw new DuplicateTypeCodeException(request.getTypeCode());
            }
            type.setTypeCode(request.getTypeCode());
        }

        if (request.getTypeName() != null) {
            type.setTypeName(request.getTypeName());
        }

        if (request.getDescription() != null) {
            type.setDescription(request.getDescription());
        }

        if (request.getRequiresBalance() != null) {
            type.setRequiresBalance(request.getRequiresBalance());
        }

        if (request.getDefaultDaysPerYear() != null) {
            type.setDefaultDaysPerYear(request.getDefaultDaysPerYear());
        }

        if (request.getIsPaid() != null) {
            type.setIsPaid(request.getIsPaid());
        }

        if (request.getRequiresApproval() != null) {
            type.setRequiresApproval(request.getRequiresApproval());
        }

        if (request.getIsActive() != null) {
            type.setIsActive(request.getIsActive());
        }

        // Validate logic AFTER all updates: requiresBalance và defaultDaysPerYear phải match
        validateBalanceAndDefaultDays(type.getRequiresBalance(), type.getDefaultDaysPerYear());

        TimeOffType updated = typeRepository.save(type);
        log.info("Updated time-off type: {}", typeId);

        return typeMapper.toResponse(updated);
    }

    /**
     * Helper method: Validate requiresBalance và defaultDaysPerYear logic
     * 
     * Business Rules:
     * - requiresBalance = true  → defaultDaysPerYear PHẢI có giá trị (để dùng cho annual reset)
     * - requiresBalance = false → defaultDaysPerYear PHẢI null (vì không cần balance tracking)
     */
    private void validateBalanceAndDefaultDays(Boolean requiresBalance, Double defaultDaysPerYear) {
        // Case 1: requiresBalance = true VÀ defaultDaysPerYear = null
        if (Boolean.TRUE.equals(requiresBalance) && defaultDaysPerYear == null) {
            throw new BadRequestAlertException(
                "Loại nghỉ phép cần balance tracking (requiresBalance = true) PHẢI có defaultDaysPerYear để sử dụng cho annual reset. " +
                "Vui lòng set defaultDaysPerYear (ví dụ: 12.0 cho 12 ngày phép/năm).",
                "TimeOffType",
                "MISSING_DEFAULT_DAYS"
            );
        }

        // Case 2: requiresBalance = false VÀ defaultDaysPerYear != null
        if (Boolean.FALSE.equals(requiresBalance) && defaultDaysPerYear != null) {
            throw new BadRequestAlertException(
                "Loại nghỉ phép không cần balance tracking (requiresBalance = false) KHÔNG thể có defaultDaysPerYear. " +
                "Field defaultDaysPerYear chỉ dùng cho các loại nghỉ phép cần check số dư (requiresBalance = true). " +
                "Vui lòng set defaultDaysPerYear = null.",
                "TimeOffType",
                "INVALID_DEFAULT_DAYS"
            );
        }
    }

    /**
     * DELETE /api/v1/admin/time-off-types/{typeId}
     * Vô hiệu hóa/Kích hoạt lại loại nghỉ phép (soft delete)
     */
    @Transactional
    public TimeOffTypeResponse toggleTimeOffTypeActive(String typeId) {
        log.debug("Request to toggle time-off type active status: {}", typeId);

        TimeOffType type = typeRepository.findById(typeId)
                .orElseThrow(() -> new TimeOffTypeNotFoundException(typeId));

        // If deactivating, check for pending requests
        if (type.getIsActive()) {
            boolean hasPendingRequests = requestRepository
                    .existsByTimeOffTypeIdAndStatus(typeId, TimeOffStatus.PENDING);

            if (hasPendingRequests) {
                throw new TimeOffTypeInUseException(typeId);
            }
        }

        // Toggle status
        type.setIsActive(!type.getIsActive());
        TimeOffType updated = typeRepository.save(type);

        log.info("Toggled time-off type {} to is_active={}", typeId, updated.getIsActive());

        return typeMapper.toResponse(updated);
    }
}
